File: apps/frontend/src/types/marketing.ts
/** Tipos do módulo Email Marketing */

// ── Templates ──────────────────────────────────────────────────────────────────

export type TemplateCategory =
    | "reengajamento"
    | "comunicado"
    | "check-up"
    | "nps"
    | "boas-vindas"
    | "newsletter"
    | "custom";

export interface EmailTemplate {
    id: string;
    organization_id: string;
    name: string;
    category: string;
    subject_template: string | null;
    html_content: string | null;
    blocks_json: string | null;
    variables: string[] | null;
    is_starter: boolean;
    created_by: string | null;
    created_at: string;
    updated_at: string;
}

export interface EmailTemplateCreateInput {
    name: string;
    subject_template?: string | null;
    html_content?: string | null;
    blocks_json?: string | null;
    variables?: string[];
    category?: string;
}

export interface EmailTemplateUpdateInput {
    name?: string;
    subject_template?: string | null;
    html_content?: string | null;
    blocks_json?: string | null;
    variables?: string[];
    category?: string;
}

export interface EmailTemplateListResponse {
    items: EmailTemplate[];
    total: number;
    limit: number;
    offset: number;
}

// ── Listas / Segmentos ─────────────────────────────────────────────────────────

export type ListType = "static" | "dynamic";

export interface EmailList {
    id: string;
    organization_id: string;
    name: string;
    description: string | null;
    list_type: string;
    filter_criteria: Record<string, unknown> | null;
    subscriber_count: number;
    created_by: string | null;
    created_at: string;
    updated_at: string;
}

export interface EmailListCreateInput {
    name: string;
    description?: string | null;
    list_type?: string;
    filter_criteria?: Record<string, unknown> | null;
}

export interface EmailListListResponse {
    items: EmailList[];
    total: number;
    limit: number;
    offset: number;
}

// ── Subscribers ────────────────────────────────────────────────────────────────

export type SubscriberStatus = "active" | "unsubscribed" | "bounced" | "complained";

export interface EmailSubscriber {
    id: string;
    organization_id: string;
    email: string;
    first_name: string | null;
    last_name: string | null;
    phone: string | null;
    patient_id: string | null;
    status: SubscriberStatus;
    tags: string[];
    custom_fields: Record<string, unknown>;
    subscribed_at: string;
    unsubscribed_at: string | null;
    created_at: string;
    updated_at: string;
}

export interface SubscriberListResponse {
    items: EmailSubscriber[];
    total: number;
    limit: number;
    offset: number;
}

// ── Campanhas ──────────────────────────────────────────────────────────────────

export type CampaignStatus =
    | "draft"
    | "scheduled"
    | "sending"
    | "sent"
    | "partial"
    | "failed"
    | "cancelled";

export interface EmailCampaign {
    id: string;
    organization_id: string;
    name: string;
    subject: string;
    template_id: string | null;
    list_id: string | null;
    from_email: string | null;
    from_name: string | null;
    reply_to: string | null;
    status: CampaignStatus;
    scheduled_at: string | null;
    sent_at: string | null;
    total_recipients: number;
    total_sent: number;
    total_delivered: number;
    total_opened: number;
    total_clicked: number;
    total_bounced: number;
    total_unsubscribed: number;
    created_by: string | null;
    created_at: string;
    updated_at: string;
}

export interface CampaignCreateInput {
    name: string;
    subject: string;
    template_id?: string | null;
    list_id?: string | null;
    from_email?: string | null;
    from_name?: string | null;
    reply_to?: string | null;
    scheduled_at?: string | null;
}

export interface CampaignUpdateInput {
    name?: string;
    subject?: string;
    template_id?: string | null;
    list_id?: string | null;
    from_email?: string | null;
    from_name?: string | null;
    reply_to?: string | null;
    scheduled_at?: string | null;
}

export interface CampaignListResponse {
    items: EmailCampaign[];
    total: number;
    limit: number;
    offset: number;
}

// ── Sends ──────────────────────────────────────────────────────────────────────

export type SendStatus =
    | "pending"
    | "sent"
    | "delivered"
    | "opened"
    | "clicked"
    | "bounced"
    | "failed"
    | "complained";

export interface EmailSend {
    id: string;
    campaign_id: string;
    subscriber_id: string;
    esp_message_id: string | null;
    status: SendStatus;
    sent_at: string | null;
    delivered_at: string | null;
    opened_at: string | null;
    clicked_at: string | null;
    bounced_at: string | null;
    error_message: string | null;
    retry_count: number;
    created_at: string;
}

// ── Métricas ───────────────────────────────────────────────────────────────────

export interface CampaignMetrics {
    campaign_id: string;
    total_recipients: number;
    total_sent: number;
    total_delivered: number;
    total_opened: number;
    total_clicked: number;
    total_bounced: number;
    total_unsubscribed: number;
    delivery_rate: number;
    open_rate: number;
    click_rate: number;
    bounce_rate: number;
    unsubscribe_rate: number;
}

export interface OverviewMetrics {
    total_campaigns: number;
    total_sent: number;
    avg_open_rate: number;
    avg_click_rate: number;
    avg_bounce_rate: number;
    recent_campaigns: EmailCampaign[];
}

// ── Domínios ───────────────────────────────────────────────────────────────────

export type DomainStatus = "pending" | "verified" | "failed";

export type TlsMode = "opportunistic" | "enforced";

export interface DnsRecord {
    record_type: string;
    name: string;
    content: string;
    ttl: string;
    priority: number | null;
    status: "pending" | "verified";
}

export interface EmailDomain {
    id: string;
    organization_id: string;
    domain: string;
    status: DomainStatus;
    region: string;
    provider: string | null;
    click_tracking: boolean;
    open_tracking: boolean;
    tls_mode: TlsMode;
    dns_records: DnsRecord[] | null;
    verified_at: string | null;
    created_by: string | null;
    created_at: string;
    updated_at: string;
}

export interface DomainCreateInput {
    domain: string;
    region?: string;
}

export interface DomainUpdateInput {
    click_tracking?: boolean;
    open_tracking?: boolean;
    tls_mode?: TlsMode;
}

export interface DomainListResponse {
    items: EmailDomain[];
    total: number;
    limit: number;
    offset: number;
}

File: apps/backend/src/schemas/marketing.py
"""Schemas Pydantic para API de Email Marketing (request/response)."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ── Templates ────────────────────────────────────────────────────────────────


class TemplateCreate(BaseModel):
    """Payload para criar template."""
    name: str = Field(..., min_length=2, max_length=255)
    category: str = Field("geral", max_length=50)
    subject_template: Optional[str] = Field(None, max_length=500)
    html_content: Optional[str] = None
    blocks_json: Optional[str] = None
    variables: Optional[list[str]] = None


class TemplateUpdate(BaseModel):
    """Payload para atualizar template (parcial)."""
    name: Optional[str] = Field(None, min_length=2, max_length=255)
    category: Optional[str] = Field(None, max_length=50)
    subject_template: Optional[str] = Field(None, max_length=500)
    html_content: Optional[str] = None
    blocks_json: Optional[str] = None
    variables: Optional[list[str]] = None


class TemplateResponse(BaseModel):
    id: str
    organization_id: str
    name: str
    category: str
    subject_template: Optional[str] = None
    html_content: Optional[str] = None
    blocks_json: Optional[str] = None
    variables: Optional[list[str]] = None
    is_starter: bool = False
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class TemplateListResponse(BaseModel):
    items: list[TemplateResponse]
    total: int
    limit: int
    offset: int


class TemplatePreviewRequest(BaseModel):
    """Payload para preview de template."""
    variables: dict[str, str] = Field(default_factory=dict)


class TemplatePreviewResponse(BaseModel):
    """Resultado do preview renderizado."""
    subject: str
    html: str
    variables_used: list[str]


# ── Listas ───────────────────────────────────────────────────────────────────


class ListCreate(BaseModel):
    """Payload para criar lista."""
    name: str = Field(..., min_length=2, max_length=255)
    description: Optional[str] = None
    list_type: str = Field("static", pattern="^(static|dynamic)$")
    filter_criteria: Optional[dict] = None


class ListUpdate(BaseModel):
    """Payload para atualizar lista (parcial)."""
    name: Optional[str] = Field(None, min_length=2, max_length=255)
    description: Optional[str] = None
    filter_criteria: Optional[dict] = None


class ListResponse(BaseModel):
    id: str
    organization_id: str
    name: str
    description: Optional[str] = None
    list_type: str
    filter_criteria: Optional[dict] = None
    subscriber_count: int = 0
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class ListListResponse(BaseModel):
    items: list[ListResponse]
    total: int
    limit: int
    offset: int


# ── Subscribers ──────────────────────────────────────────────────────────────


class SubscriberCreate(BaseModel):
    email: str = Field(..., max_length=255)
    name: Optional[str] = Field(None, max_length=255)
    patient_id: Optional[str] = None


class SubscriberResponse(BaseModel):
    id: str
    organization_id: str
    email: str
    name: Optional[str] = None
    patient_id: Optional[str] = None
    status: str
    unsubscribed_at: Optional[datetime] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class CsvImportResponse(BaseModel):
    """Resultado de importacao CSV."""
    total_rows: int
    created: int
    skipped: int
    errors: list[str]


class PatientSyncResponse(BaseModel):
    """Resultado do sync Patient→Subscriber."""
    synced: int
    unsubscribed: int
    skipped: int


# ── Campanhas ────────────────────────────────────────────────────────────────


class CampaignCreate(BaseModel):
    """Payload para criar campanha."""
    name: str = Field(..., min_length=2, max_length=255)
    subject: str = Field(..., min_length=2, max_length=500)
    template_id: Optional[str] = None
    list_id: Optional[str] = None
    from_email: Optional[str] = Field(None, max_length=255)
    from_name: Optional[str] = Field(None, max_length=255)
    reply_to: Optional[str] = Field(None, max_length=255)
    scheduled_at: Optional[datetime] = None


class CampaignUpdate(BaseModel):
    """Payload para atualizar campanha (parcial)."""
    name: Optional[str] = Field(None, min_length=2, max_length=255)
    subject: Optional[str] = Field(None, min_length=2, max_length=500)
    template_id: Optional[str] = None
    list_id: Optional[str] = None
    from_email: Optional[str] = Field(None, max_length=255)
    from_name: Optional[str] = Field(None, max_length=255)
    reply_to: Optional[str] = Field(None, max_length=255)
    scheduled_at: Optional[datetime] = None


class CampaignResponse(BaseModel):
    id: str
    organization_id: str
    name: str
    subject: str
    template_id: Optional[str] = None
    list_id: Optional[str] = None
    from_email: Optional[str] = None
    from_name: Optional[str] = None
    reply_to: Optional[str] = None
    status: str
    scheduled_at: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    total_recipients: int = 0
    total_sent: int = 0
    total_delivered: int = 0
    total_opened: int = 0
    total_clicked: int = 0
    total_bounced: int = 0
    total_unsubscribed: int = 0
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class CampaignScheduleRequest(BaseModel):
    """Payload para agendar campanha."""
    scheduled_at: datetime


class CampaignListResponse(BaseModel):
    items: list[CampaignResponse]
    total: int
    limit: int
    offset: int


# ── Métricas ─────────────────────────────────────────────────────────────────


class CampaignMetricsResponse(BaseModel):
    """Metricas detalhadas de uma campanha."""
    campaign_id: str
    total_recipients: int = 0
    total_sent: int = 0
    total_delivered: int = 0
    total_opened: int = 0
    total_clicked: int = 0
    total_bounced: int = 0
    total_unsubscribed: int = 0
    delivery_rate: float = 0.0
    open_rate: float = 0.0
    click_rate: float = 0.0
    bounce_rate: float = 0.0
    unsubscribe_rate: float = 0.0


class OverviewMetricsResponse(BaseModel):
    """Metricas globais do modulo de marketing."""
    total_campaigns: int = 0
    total_sent: int = 0
    avg_open_rate: float = 0.0
    avg_click_rate: float = 0.0
    avg_bounce_rate: float = 0.0
    recent_campaigns: list[CampaignResponse] = []


# ── Domínios ─────────────────────────────────────────────────────────────────


class DnsRecordResponse(BaseModel):
    """Registro DNS individual."""
    record_type: str  # TXT, MX, CNAME
    name: str
    content: str
    ttl: str = "Auto"
    priority: Optional[int] = None
    status: str = "pending"  # pending, verified


class DomainCreate(BaseModel):
    """Payload para adicionar domínio."""
    domain: str = Field(..., min_length=3, max_length=255)
    region: str = Field("sa-east-1", max_length=30)


class DomainUpdate(BaseModel):
    """Payload para atualizar configuração do domínio."""
    click_tracking: Optional[bool] = None
    open_tracking: Optional[bool] = None
    tls_mode: Optional[str] = Field(None, pattern="^(opportunistic|enforced)$")


class DomainResponse(BaseModel):
    id: str
    organization_id: str
    domain: str
    status: str
    region: str
    provider: Optional[str] = None
    click_tracking: bool = True
    open_tracking: bool = False
    tls_mode: str = "opportunistic"
    dns_records: Optional[list[DnsRecordResponse]] = None
    verified_at: Optional[datetime] = None
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class DomainListResponse(BaseModel):
    items: list[DomainResponse]
    total: int
    limit: int
    offset: int


# ── Inbound Emails ───────────────────────────────────────────────────────────


class InboundEmailResponse(BaseModel):
    """Resposta para consulta de emails inbound."""
    id: str
    organization_id: str
    from_email: str
    from_name: Optional[str] = None
    to_email: str
    subject: Optional[str] = None
    text_content: Optional[str] = None
    html_content: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime
    model_config = ConfigDict(from_attributes=True)


class InboundEmailListResponse(BaseModel):
    """Lista de emails inbound."""
    items: list[InboundEmailResponse]
    total: int
    limit: int
    offset: int


class SingleEmailRequest(BaseModel):
    """Payload para envio de email unico."""
    to_email: str = Field(..., max_length=255)
    subject: str = Field(..., min_length=2, max_length=500)
    html_content: Optional[str] = None
    text_content: Optional[str] = None
    from_name: Optional[str] = Field(None, max_length=255)
    from_email: Optional[str] = Field(None, max_length=255)

